proposals/lip_define_state_and_state_transitions_of_dpos_module.md

```diff
@@ -0,0 +1,1481 @@
+        "name": validator.name,
+        "totalVotesReceived": totalVotesReceived,
+        "selfVotes": selfVotes,
+        "lastGeneratedHeight": validator.lastGeneratedHeight,
+        "isBanned": validator.isBanned,
+        "pomHeights": validator.pomHeight,
+        "consecutiveMissedBlocks": validator.consecutiveMissedBlocks
+    } serialized using delegateStoreSchema.
+    ```
+    Further, for every entry `validator` in `genesisBlockAssetObject.validators`, also create an entry in the name substore with 
+    ```python
+    storeKey = validator.name utf8-encoded
+    storeValue = {
+        "delegateAddress": validator.address
+    } serialized using nameStoreSchema.
+    ```
+
+* For each entry `voter` in `genesisBlockAssetObject.voters`, create an entry in the voter substore  with
+    ```python
+    storeKey = uint32be(voter.address)
+    storeValue = {
+        "sentVotes": voter.sentVotes,
+        "pendingUnlocks": voter.pendingUnlocks
+    } serialized using voterStoreSchema.
+    ```
+
+* For each entry `snapshot` in `genesisBlockAssetObject.snapshots`, create an entry in the snapshot substore  with
+    ```python
+    storeKey = uint32be(snapshot.roundNumber)
+    storeValue = {
+        "activeDelegates": snapshot.activeDelegates,
+        "delegateWeightSnapshot": snapshot.delegateWeightSnapshot
+    } serialized using snapshotStoreSchema.
+    ```
+
+* Create an entry in the genesis data substore with
+    ```python
+    storeKey = EMPTY_BYTES
+    storeValue = {
+        "height": block header height of the genesis block,
+        "initRounds": genesisBlockAssetObject.genesisData.initRounds,
+        "initDelegates": genesisBlockAssetObject.genesisData.initDelegates
+    } serialized using genesisDataStoreSchema.
+    ```
+
+#### Genesis State Finalization
+
+To finalize the state of the genesis block the following logic is executed.  If any step fails, the block is discarded and has no further effect.
+
+As in the previous point, let `genesisBlockAssetBytes` be the `data` bytes included in the block assets for the DPoS module and let `genesisBlockAssetObject` be the deserialization of `genesisBlockAssetBytes` according to the `genesisDPoSStoreSchema` schema, given above. Then:
+
+```python
+# register all validators in the Validators module
+for validator in genesisBlockAssetObject.validators:
+    validators.registerValidatorKeys(
+        validator.address,
+        validator.proofOfPossession,
+        validator.generatorKey,
+        validator.blsKey
+    )
+    if the above returns False:
+        fail
+
+# check that all sentVotes and pendingUnlocks correspond to locked tokens
+for address a key of the voter substore:
+    votedAmount = 0
+    for sentVote in voter(address).sentVotes:
+        votedAmount += sentVote.amount
+    for pendingUnlock in voter(address).pendingUnlocks:
+        votedAmount += pendingUnlock.amount
+
+    if token.getLockedAmount(address, MODULE_ID_DPOS, TOKEN_ID_DPOS) != votedAmount:
```

Comment:
for validator/bft, it is actively initializing, but for token module, it expects token module to initialize. Is it intentional?

Reply:
yes, we did it this way intentionally, but we had arguments going both ways. One of the reasons for doing this way was: 
- in Token module, you can set locked tokens in the genesis state. So if we directly call `lock` here, you run into the scenario where the genesis state could lock tokens twice, once in Toke module directly, once from here. This is not possible in Validators or BFT, because they always start with an empty state.
- this also allows us to check for equality, and hence be sure that exactly the right amount of tokens are locked. 
- I found it more natural when checking the genesis state that you would see the locked tokens, and not see only available balance (which gets locked during processing).

If you think we should go for the other way, it would also be fine for me (it was also Alessandro's preference). I would just change the name of the function here.